resources/js/components/felte-form/FelteForm.tsx

```diff
@@ -0,0 +1,104 @@
+import {
+  AssignableErrors,
+  FormConfigWithoutTransformFn,
+  Obj,
+} from "@felte/core";
+import { reporter } from "@felte/reporter-solid";
+import { createForm } from "@felte/solid";
+import { validator } from "@felte/validator-zod";
+import { useTransContext } from "@mbarzda/solid-i18next";
+import { isAxiosError } from "axios";
+import { Api } from "data-access/memo-api/types";
+import { Context, JSX, createContext, splitProps, useContext } from "solid-js";
+import toast from "solid-toast";
+import { ZodSchema } from "zod";
+
+type FormContextValue<T extends Obj> = {
+  props: FormProps<T>;
+  form: ReturnType<typeof createForm<T>>;
+};
+
+const FormContext = createContext(undefined, {
+  name: "FormContext",
+});
+
+type FormProps<T extends Obj> = Omit<
+  JSX.FormHTMLAttributes<HTMLFormElement>,
+  "onSubmit" | "onError"
+> &
+  FormConfigWithoutTransformFn<T> & {
+    schema: ZodSchema<T>;
+  };
+
+/**
+ * Wrapper for felte's `createForm`
+ *
+ * Includes solidjs' Provider, that stores
+ * createForm's data and component props
+ */
+export const FelteForm = <T extends Obj>(props: FormProps<T>) => {
+  const [t] = useTransContext();
+  const [local, createFormOptions, others] = splitProps(
+    props,
+    ["children", "schema"],
+    [
+      "debounced",
+      "extend",
+      "initialValues",
+      "onError",
+      "onSubmit",
+      "onSuccess",
+      "transform",
+      "validate",
+      "warn",
+    ]
+  );
+
+  const form = createForm<T>({
+    ...createFormOptions,
+    extend: [validator({ schema: local.schema }), reporter],
+    onError: (error, ctx) => {
+      createFormOptions?.onError?.(error, ctx);
+      if (isAxiosError<Api.ErrorResponse>(error)) {
+        const validationErrors = error.response?.data.errors.reduce(
+          (prev, curr) => {
+            if (Api.isValidationError(curr)) {
+              // @ts-ignore
+              prev[curr.field] = t(curr.code, {
+                attribute: t(curr.field),
+                ...curr.data,
+              });
+            }
+            return prev;
+          },
+          {} as AssignableErrors<T>
+        );
+        if (validationErrors) ctx.setErrors(validationErrors);
+        error.response?.data.errors.filter(Api.isBaseError).forEach((error) => {
+          toast.error(t(error.code));
+        });
+      }
+    },
+  });
+
+  return (
+    <FormContext.Provider value={{ form, props }}>
+      <form ref={form.form} {...others}>
+        <fieldset class="contents" disabled={form.isSubmitting()}>
+          {local.children}
+        </fieldset>
+      </form>
+    </FormContext.Provider>
+  );
+};
+
+export const useFormContext = <T extends Obj>() => {
+  const value = useContext(
+    FormContext as unknown as Context<FormContextValue<T>>
```

Comment:
Dlaczego to musi być takie brzydkie? Może jakieś małe wyjaśnienie w komentarzu, ogólnie w tym pliku już od góry trochę nie wiem co się dzieje, przydałoby się krótkie wyjaśnienie ogólne mechanizmu i co my tu robimy.

Reply:
ok, napisze docsy, ale tutaj też wytłumaczę. Generalnie ani react, ani solid nie dają ładnego sposobu na generic context. To jest jeden z sposobów na stworzenie takiego contextu. Zależało mi na tym, bo przy dużych formularzach to się przyda.

```ts
const formContext = useFormContext<{ field1: string; field2: number}>();
// formContext.form <- ładnie otypowany obiekt tak jak jest zwracany z createForm 
```

Reply:
i co z tym?

Reply:
No więc szczerze powiem że to co napisałeś w komentarzu to tak mniej więcej rozumiem (chociaż nie rozumiem tak bez wnikania jak to tak naprawdę działa), natomiast te komentarze w kodzie są mało pomocne jak dla mnie :) Więc jak ci się chce to opisz to jeszcze trochę dokładniej w komentarzu gdzieś na górze pliku (nie JSDoc tylko tak luzem). A jeśli nie masz weny to ujdzie jak jest, resolvuj.

Reply:
pewnie jak się pojawi bardziej skomplikowany formularz to zauważysz o co chodzi